src/ServiceControl/Monitoring/HeartbeatMonitor.cs

```diff
@@ -0,0 +1,53 @@
+namespace ServiceControl.Monitoring
+{
+    using System;
+    using System.Threading;
+
+    class HeartbeatMonitor
+    {
+        volatile RecordedHeartbeat heartbeat = new RecordedHeartbeat(HeartbeatStatus.Unknown, null);
+
+        public void MarkAlive(DateTime timestamp)
+        {
+            var newReading = new RecordedHeartbeat(HeartbeatStatus.Alive, timestamp);
+            var done = false;
+
+            while (!done)
+            {
+                var priorReading = heartbeat;
+
+                if (priorReading.Timestamp.GetValueOrDefault() <= timestamp)
+                {
+                    done = Interlocked.CompareExchange(ref heartbeat, newReading, priorReading) == priorReading;
+                }
+                else
+                {
+                    done = true;
+                }
+            }
+        }
+
+        public RecordedHeartbeat Age(DateTime cutoff)
```

Comment:
How about calling this `MarkDeadIfNoRecentHeartbeat` ? Every time I come back to this class I need to figure out what does the `Age` method do ;-)